File: app/repositories/user_anime_repository.py
from sqlalchemy.orm import Session
from sqlalchemy import select

from app.models.user_anime import UserAnime
from app.models.anime import Anime


class UserAnimeRepository:

    def __init__(self, db: Session):
        self.db = db

    def get_by_id(self, entry_id: int) -> UserAnime | None:
        return self.db.scalar(
            select(UserAnime).where(UserAnime.id == entry_id)
        )

    def get_by_user_and_anime(self, user_id: int, anime_id: int) -> UserAnime | None:
        return self.db.scalar(
            select(UserAnime).where(
                UserAnime.user_id == user_id,
                UserAnime.anime_id == anime_id,
            )
        )

    def get_all_by_user(self, user_id: int) -> list[dict]:
        # Faz join e retorna dicts com dados dos dois modelos
        rows = self.db.execute(
            select(UserAnime, Anime)
            .join(Anime, Anime.id == UserAnime.anime_id)
            .where(UserAnime.user_id == user_id)
        ).all()

        result = []
        for user_anime, anime in rows:
            result.append({
                "id": user_anime.id,
                "anime_id": user_anime.anime_id,
                "user_id": user_anime.user_id,
                "status": user_anime.status,
                "rating": user_anime.rating,
                "is_favorite": user_anime.is_favorite,
                "added_at": user_anime.added_at,
                "anime": {
                    "id": anime.id,
                    "anilist_id": anime.anilist_id,
                    "title_romaji": anime.title_romaji,
                    "title_english": anime.title_english,
                    "cover_image_url": anime.cover_image_url,
                    "episode_count": anime.episode_count,
                    "genres": anime.genres,
                }
            })
        return result

    def save(self, entry: UserAnime) -> UserAnime:
        self.db.add(entry)
        self.db.commit()
        self.db.refresh(entry)
        return entry

    def delete(self, entry: UserAnime) -> None:
        self.db.delete(entry)
        self.db.commit()

File: backend-rust/src/repositories/user_anime_repository.rs
use chrono::NaiveDateTime;
use sqlx::PgPool;

use crate::models::user_anime::{UserAnime, UserAnimeWithAnime};

pub async fn anime_exists(db: &PgPool, anime_id: i32) -> Result<bool, sqlx::Error> {
    let exists = sqlx::query_scalar::<_, bool>(
        r#"
        SELECT EXISTS (
            SELECT 1
            FROM anime_cache
            WHERE id = $1
        )
        "#,
    )
    .bind(anime_id)
    .fetch_one(db)
    .await?;

    Ok(exists)
}

pub async fn find_by_user_and_anime(
    db: &PgPool,
    user_id: i32,
    anime_id: i32,
) -> Result<Option<UserAnime>, sqlx::Error> {
    sqlx::query_as::<_, UserAnime>(
        r#"
        SELECT id, user_id, anime_id, status, rating, is_favorite, added_at
        FROM user_anime
        WHERE user_id = $1 AND anime_id = $2
        "#,
    )
    .bind(user_id)
    .bind(anime_id)
    .fetch_optional(db)
    .await
}

pub async fn list_by_user(
    db: &PgPool,
    user_id: i32,
) -> Result<Vec<UserAnimeWithAnime>, sqlx::Error> {
    sqlx::query_as::<_, UserAnimeWithAnime>(
        r#"
        SELECT
            ua.id,
            ua.user_id,
            ua.anime_id,
            ua.status,
            ua.rating,
            ua.is_favorite,
            ua.added_at,

            a.anilist_id AS anime_anilist_id,
            a.title_romaji AS anime_title_romaji,
            a.title_english AS anime_title_english,
            a.cover_image_url AS anime_cover_image_url,
            a.episode_count AS anime_episode_count,
            a.genres AS anime_genres
        FROM user_anime ua
        JOIN anime_cache a ON a.id = ua.anime_id
        WHERE ua.user_id = $1
        ORDER BY ua.added_at DESC, ua.id DESC
        "#,
    )
    .bind(user_id)
    .fetch_all(db)
    .await
}

pub async fn create(
    db: &PgPool,
    user_id: i32,
    anime_id: i32,
    status: &str,
    rating: Option<i32>,
    is_favorite: bool,
    added_at: NaiveDateTime,
) -> Result<UserAnime, sqlx::Error> {
    sqlx::query_as::<_, UserAnime>(
        r#"
        INSERT INTO user_anime (
            user_id,
            anime_id,
            status,
            rating,
            is_favorite,
            added_at
        )
        VALUES ($1, $2, $3, $4, $5, $6)
        RETURNING id, user_id, anime_id, status, rating, is_favorite, added_at
        "#,
    )
    .bind(user_id)
    .bind(anime_id)
    .bind(status)
    .bind(rating)
    .bind(is_favorite)
    .bind(added_at)
    .fetch_one(db)
    .await
}

pub async fn update(
    db: &PgPool,
    id: i32,
    status: &str,
    rating: Option<i32>,
    is_favorite: bool,
) -> Result<UserAnime, sqlx::Error> {
    sqlx::query_as::<_, UserAnime>(
        r#"
        UPDATE user_anime
        SET
            status = $1,
            rating = $2,
            is_favorite = $3
        WHERE id = $4
        RETURNING id, user_id, anime_id, status, rating, is_favorite, added_at
        "#,
    )
    .bind(status)
    .bind(rating)
    .bind(is_favorite)
    .bind(id)
    .fetch_one(db)
    .await
}

pub async fn delete_by_id(db: &PgPool, id: i32) -> Result<(), sqlx::Error> {
    sqlx::query(
        r#"
        DELETE FROM user_anime
        WHERE id = $1
        "#,
    )
    .bind(id)
    .execute(db)
    .await?;

    Ok(())
}
